Fix building flat roofs with user provided data when terrain is enabled

karimnaaji opened the pull request:
This is a small change that removes the dependency on the feature data having the property named 'height' in its data fields and adds render tests and release testing coverage for this. This enables flat roofs always, to allow custom data source to use the code path for flat roofs, as explained in #10170 . There are valid use cases to conditionally enable flat roofs using a style spec property for explicit usage, we will evaluate independently of this fix (refer https://github.com/mapbox/mapbox-gl-js/issues/10170#issuecomment-770496123). The render tests and demo are using a rather large example (taken from the user demo provided in the ticket), let me know if this needs to be trimmed down a bit.

Using the custom data source from https://github.com/mapbox/mapbox-gl-js/issues/10170#issuecomment-742918373, added from demo/fill-extrusion-querying:

**Before**
<img width="587" alt="Screen Shot 2021-02-01 at 2 47 25 PM" src="https://user-images.githubusercontent.com/7061573/106527635-74ab6f00-649c-11eb-95ad-715e1055e36f.png">

**After**
<img width="585" alt=" " src="https://user-images.githubusercontent.com/7061573/106527627-72e1ab80-649c-11eb-9673-4ee2c4689367.png">

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] tagged @mapbox/gl-native if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Improve terrain support of fill-extrusions flat roofs when used with custom data sources</changelog>`

mourner commented on src/data/bucket/fill_extrusion_bucket.js:
```diff
@@ -301,7 +301,7 @@ class FillExtrusionBucket implements Bucket {
 
     addFeature(feature: BucketFeature, geometry: Array<Array<Point>>, index: number, canonical: CanonicalTileID, imagePositions: {[_: string]: ImagePosition}) {
         const flatRoof = this.enableTerrain && feature.properties && feature.properties.hasOwnProperty('type') &&
```
I suppose we don't need the `type` check in `properties` as well? It's different from `feature.type` and also a data property we don't want to rely on.

```suggestion
        const flatRoof = this.enableTerrain && feature.properties &&
```

astojilj replied:
Current code is a conservative hack (best effort) to enable flat roofs for some of the buildings.
Don't think it is a good idea to enable it for any feature that is larger than the building (countries, routes) - those kind of features would span across multiple tiles and we wouldn't be able to control when flat roof is used and where it is not.
> There are valid use cases to conditionally enable flat roofs using a style spec property for explicit usage, we will evaluate independently of this fix (refer #10170 (comment))

IMHO, that style spec property for explicit usage would be a great fix for this, too - I'd be concerned about regressions if merging this PR without style spec property for explicit usage.

mourner commented on src/render/draw_fill_extrusion.js:
```diff
@@ -138,6 +138,13 @@ function flatRoofsUpdate(context, source, coord, bucket, layer, terrain) {
+                const nBucket: ?FillExtrusionBucket = (n.getBucket(layer): any);
+                if (nBucket) return nBucket;
+            }
+        }
```
Now that we have to repeat this chunk of code, perhaps there's a way to refactor the `getLoadedBucket` function for simplicity, e.g. cutting this bit out as a separate function and getting rid of the loop later?

karimnaaji replied:
Addressed in a9d4c16a47b90dd464a93c2cefd6c0c4d1bbc2b0.

mourner reviewed (changes requested):
> The render tests and demo are using a rather large example (taken from the user demo provided in the ticket), let me know if this needs to be trimmed down a bit.

Yeah, +27k lines of code just for a small fix is too much to add. I would try to make a much simpler test / debug page.

karimnaaji commented:
Addressed comments and updates to further control the behavior in https://github.com/mapbox/mapbox-gl-js/pull/10347/commits/79361bc8b915e7454a9f730e6a075783765cdb55: 
- Added style spec property `fill-extrusion-flat-roofs` to allow controlling flat roofs as suggested in https://github.com/mapbox/mapbox-gl-js/pull/10347#discussion_r568484982 and https://github.com/mapbox/mapbox-gl-js/issues/10170#issuecomment-770496123, the spec is the following, defaulting to `true` to allow for backward compatibility:
```
    "fill-extrusion-flat-roofs": {
      "type": "boolean",
      "default": true,
      "doc": "Whether or not the extruded fill should use flat roofs when used along with terrain.",
      "sdk-support": {
        "basic functionality": {
          "js": "2.2.0"
        }
      },
      "expression": {
        "interpolated": false,
        "parameters": [
          "zoom"
        ]
      },
      "property-type": "data-constant"
    },
```
cc @mapbox/map-design-team for review of the spec.

- Updated fill-extrusion querying to account for this spec:

https://user-images.githubusercontent.com/7061573/106676557-00d89780-656c-11eb-99c1-50c532988da5.mov

- Added tests for the above and trimmed down other regression and example sizes

@mourner @astojilj PTAL.

asheemmamoowala reviewed (changes requested):
The introduction of this new style spec property changes the default behavior of `fill-extrusion` layers  and breaks compatibility with existing styles with terrain enabled. Would you consider flattening extrusions on terrain by default as the pre-dominant use might be with smaller polygons? For polygons where the use is not expected to auto adjust the extrusion height against terrain, would the default behavior be to simply slope along side terrain or to remove the terrain adjustment and set height based on mean sea level? Is there enough of a use case to directly support extrusions that slope?

The `fill-extrusion-flat-roofs` name seems very specific to buildings, ie: roofs. Is there a more generic name that can be used here?

karimnaaji commented:
>The introduction of this new style spec property changes the default behavior of fill-extrusion layers and breaks compatibility with existing styles with terrain enabled.

The default behavior is the current behavior (flattening) to explicitly not break compatibility, do you have any specific example that could help clarify this concern?

>For polygons where the use is not expected to auto adjust the extrusion height against terrain, would the default behavior be to simply slope along side terrain or to remove the terrain adjustment and set height based on mean sea level? Is there enough of a use case to directly support extrusions that slope?

Currently it is set to slope along side the terrain. The main use case is described in https://github.com/mapbox/mapbox-gl-js/issues/10170#issuecomment-770737581 for vegetation coverage, but this is mainly to allow control over turning this off:

Since we don't always honor flat roofs due to the small spans area restriction (flat roofs are usually only supported on a subset of fill-extrusions which are polygons that do not span across multiple tiles), the introduction of the spec would allow to turn this off and not automatically have flattening enabled when terrain is on. Having flat roofs always, with custom user data sets that may not be fill-extrusions representing buildings, may become inconvenient if we don't allow such control, especially considering that these polygons may or may not be spanning over a few tiles.

That being said, I'm happy to revert the commits for the initially stated scope of this PR for further evaluation.

asheemmamoowala commented:
> The default behavior is the current behavior (flattening) to explicitly not break compatibility, do you have any specific example that could help clarify this concern?

Apologies for blowing past that part of the default value. 🤦 

> Since we don't always honor flat roofs due to the small spans area restriction (flat roofs are usually only supported on a subset of fill-extrusions which are polygons that do not span across multiple tiles), the introduction of the spec would allow to turn this off and not automatically have flattening enabled when terrain is on. Having flat roofs always, with custom user data sets that may not be fill-extrusions representing buildings, may become inconvenient if we don't allow such control, especially considering that these polygons may or may not be spanning over a few tiles.

So when polygons spanning multiple tiles need to be rendered, how does using this property provide a result that a map designer would want?

karimnaaji commented:
>So when polygons spanning multiple tiles need to be rendered, how does using this property provide a result that a map designer would want?

I don't think the the spec toggle has any side effect preventing further improvements over the multiple-tile spanning polygon flat roofs case, which is our current behavior and default. But in doubt, I reverted the spec addition to only the initial scope of this PR.

arindam1993 commented on src/data/bucket/fill_extrusion_bucket.js:
```diff
@@ -300,8 +300,8 @@ class FillExtrusionBucket implements Bucket {
     addFeature(feature: BucketFeature, geometry: Array<Array<Point>>, index: number, canonical: CanonicalTileID, imagePositions: {[_: string]: ImagePosition}) {
-        const flatRoof = this.enableTerrain && feature.properties && feature.properties.hasOwnProperty('type') &&
-            feature.properties.hasOwnProperty('height') && vectorTileFeatureTypes[feature.type] === 'Polygon';
+        const flatRoof = this.enableTerrain && feature.properties &&
```
I think you can get rid of `feature.properties && vectorTileFeatureTypes[feature.type] === 'Polygon'` as well,
requiring properties for flat tops is weird.
Currently we require feature-type `Polygon` as a requirement tessellating and filling out the roof. But if someone is using LineString to have `open-top` extrusions, I don't see why we should exclude them from being flat tops.

karimnaaji replied:
I think that ties to @astojilj's comment here https://github.com/mapbox/mapbox-gl-js/pull/10347#discussion_r568484982.

>Current code is a conservative hack (best effort) to enable flat roofs for some of the buildings.
Don't think it is a good idea to enable it for any feature that is larger than the building (countries, routes)

I'd be concerned about the behavior if we begin to be even less conservative here. The main issue is what has exposed #10170, we should fix that case first, and address your suggestion if needed. I'll hold on this suggestion.

arindam1993 commented on debug/fill-extrusion-querying.html:
```diff
@@ -0,0 +1,1391 @@
+<!DOCTYPE html>
```
we already have one of these called `extrusion-querying`

Why not add the new Geojson to that?

karimnaaji replied:
Nice, I wasn't aware of that one, I'll just remove this new one.

karimnaaji replied:
Updated in https://github.com/mapbox/mapbox-gl-js/pull/10347/commits/2b33470ff74c65b4b4a8960a61b2846b2ae2eaef.
